pkg/auth/oidc_connect.go

```diff
@@ -0,0 +1,294 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/allegro/bigcache"
+	"github.com/coreos/go-oidc"
+	"github.com/go-logr/logr"
+	"github.com/projectcontour/contour-authserver/pkg/config"
+	"github.com/projectcontour/contour-authserver/pkg/store"
+	"golang.org/x/oauth2"
+)
+
+const (
+	stateQueryParamName = "conauth"
+	oauthTokenName      = "contourtoken"
+)
+
+// OIDCConnect .
+type OIDCConnect struct {
+	Log        logr.Logger
+	OidcConfig *config.OIDCConfig
+	Cache      *bigcache.BigCache
+	HTTPClient *http.Client
+	provider   *oidc.Provider
+}
+
+// Implement interface.
+var _ Checker = &OIDCConnect{}
+
+//Check ...
+func (h *OIDCConnect) Check(ctx context.Context, req *Request) (*Response, error) {
+
+	h.Log.Info("checking....")
+	h.Log.Info("checking request",
+		"host", req.Request.Host,
+		"path", req.Request.URL.Path,
+		"id", req.ID,
+	)
+
+	if h.provider == nil {
+		h.provider, _ = h.initProvider(ctx)
+	}
+
+	// Generate response object.
+	resp := Response{
+		Response: http.Response{
+			StatusCode: http.StatusUnauthorized,
+			Header:     http.Header{},
+		},
+	}
+
+	h.Log.Info(formatRequest(req))
+	url := parseURL(req)
+	h.verifyRequestHandler(ctx, req, &resp, url)
+
+	return &resp, nil
+}
+
+func (h *OIDCConnect) verifyRequestHandler(ctx context.Context, req *Request, resp *Response, url *url.URL) {
```

Comment:
Nit: I would probably collapse this into the `Check` method above. I think it would be easier to read if the handlers returned the `(*Response, error)` directly.

pseudocode:
```
Check() {
    if path {
        return callbackHandler()
    }

    if !validState {
        return loginHandler()
    }

    // accept
    return &resp, nil
}
```